client/selfrepair/tests/test_utils.js

```diff
@@ -69,3 +69,101 @@ describe('stableSample', () => {
     }
   });
 });
+
+function randInt(min, max) {
+  return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+describe('bucketSample', () => {
+  it('should match at about the right rate', () => {
+    const trials = 1000;
+    const total = randInt(0, 10000);
+    const start = randInt(0, total);
+    const count = randInt(0, total - start);
+    let hits = 0;
+    for (let i = 0; i < trials; i++) {
+      if (bucketSample(Math.random(), start, count, total)) {
+        hits += 1;
+      }
+    }
+    // 95% accurate
+    const rate = count / total;
+    expect(Math.abs((hits / trials) - rate) < 0.05).toEqual(true);
+  });
+
+  it('should be stable', () => {
+    for (let i = 0; i < 100; i++) {
+      const total = randInt(0, 10000);
+      const start = randInt(0, total);
+      const count = randInt(0, total - start);
+      const val = Math.random();
+      const hit = bucketSample(val, start, count, total);
+      for (let j = 0; j < 10; j++) {
+        expect(bucketSample(val, start, count, total)).toEqual(hit);
+      }
+    }
+  });
+
+  it('should group inputs into non-overlapping buckets', () => {
+    for (let i = 0; i < 100; i++) {
+      const total = randInt(100, 10000);
+
+      // Create 10 buckets, with the end of the count as the last.
+      const bucketPoints = [];
+      for (let k = 0; k < 9; k++) {
+        bucketPoints.push(randInt(0, total));
+      }
+      bucketPoints.push(total);
+      bucketPoints.sort((a, b) => a - b);
```

Comment:
I think you can generated bucketPoints more simply like this

```javascript
const bucketCount = 10;
const bucketPoints = [0];
for (let j = 0; j < bucketCount; j++) {
  const delta = randInt(total / bucketCount / 2, total / bucketCount);
  const last = bucketPoints[bucketPoints.length - 1]
  bucketPoints.push(last + delta);
}
```

Then below when you are iterating over this, you don't have to special case 0, because it is already there, and you don't have to special case `threeInARow`.